File: intv_set_google_sol.py
# Google Interview problem in two approaches
# ***commented out print - used only for debugging purpose

#  Naive
def hasPairWithSum(arr, sum):
  arrLen = len(arr)
  for i in range(0, arrLen):
    for j in range(arrLen):
      if (arr[i] + arr[j] == sum):
        return True
  return False

# Better
def hasPairWithSum2(arr, sum):
  mySet = set()
  arrLen = len(arr)
  for i in range(0, arrLen):
    # print(i, arr[i])
    if arr[i] in mySet:
      # print(mySet)
      return True
    # print('in hash', sum - arr[i])
    mySet.add(sum - arr[i])
  return False

hasPairWithSum2([6, 4, 3, 2, 1, 7], 9)
# print(hasPairWithSum2([6, 4, 3, 2, 1, 7], 9))

File: intv_set_google_sol.js
// Naive
function hasPairWithSum(arr, sum) {
  var len = arr.length
  for (var i = 0; i < len - 1; i++) {
    for (var j = i + 1; j < len; j++) {
      if (arr[i] + arr[j] === sum) return true
    }
  }

  return false
}

// Better
function hasPairWithSum2(arr, sum) {
  const mySet = new Set()
  const len = arr.length
  for (let i = 0; i < len; i++) {
    console.log(i, arr[i])
    if (mySet.has(arr[i])) {
      console.log(mySet)
      return true
    }
    console.log('in hash', sum - arr[i])
    mySet.add(sum - arr[i])
  }
  return false
}

hasPairWithSum2([6, 4, 3, 2, 1, 7], 9)
// console.log(hasPairWithSum2([6, 4, 3, 2, 1, 7], 9))
